signer/genericrsa/rsa.go

```diff
@@ -0,0 +1,273 @@
+package genericrsa
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"hash"
+	"io"
+
+	"go.mozilla.org/autograph/formats"
+
+	"github.com/pkg/errors"
+	"go.mozilla.org/autograph/signer"
+)
+
+const (
+	// Type of this signer is "rsa"
+	Type = "genericrsa"
+)
+
+// RSASigner holds the configuration of the signer
+type RSASigner struct {
+	signer.Configuration
+
+	// key is the RSA private key to sign hashes.
+	// we use the `crypto.PrivateKey` interface to support
+	// keys in HSM.
+	key crypto.PrivateKey
+
+	// pubkey is an RSA Public Key
+	pubKey crypto.PublicKey
+
+	// rng is our random number generator
+	rng io.Reader
+
+	// sigOpts stores options for PSS signing and is nil
+	// in pkcs15 mode
+	sigOpts crypto.SignerOpts
+
+	// hashSize is the byte size of the configured hash checksum
+	hashSize int
+}
+
+const (
+	// ModePSS enables PSS padding mode
+	ModePSS = "pss"
+
+	// ModePKCS15 enables PKCS15 padding mode
+	ModePKCS15 = "pkcs15"
+)
+
+// Options contains options for creating and verifying PKCS15 signatures.
+type Options struct {
+	// Hash, if not zero, overrides the hash function passed to SignPSS.
+	// This is the only way to specify the hash function when using the
+	// crypto.Signer interface.
+	Hash crypto.Hash
+}
+
+// HashFunc returns the Hash used by the signer so that Options implements
+// crypto.SignerOpts
+func (opts *Options) HashFunc() crypto.Hash {
+	return opts.Hash
+
+}
+
+// New initializes a rsa signer using a configuration
+func New(conf signer.Configuration) (s *RSASigner, err error) {
+	s = new(RSASigner)
+
+	if conf.Type != Type {
+		return nil, errors.Errorf("genericrsa: invalid type %q, must be %q", conf.Type, Type)
+	}
+	s.Type = conf.Type
+
+	if conf.ID == "" {
+		return nil, errors.New("genericrsa: missing signer ID in signer configuration")
+	}
+	s.ID = conf.ID
+
+	switch conf.Mode {
+	case ModePSS, ModePKCS15:
+		s.Mode = conf.Mode
+	case "":
+		return nil, errors.Errorf("genericrsa: missing signer mode for signer %q, must be 'pkcs15' or 'pss'", s.ID)
+	default:
+		return nil, errors.Errorf("genericrsa: invalid signer mode %q for signer %q, must be 'pkcs15' or 'pss'", conf.Mode, s.ID)
+	}
+
+	if conf.PrivateKey == "" {
+		return nil, errors.Errorf("genericrsa: missing private key for signer %q", s.ID)
+	}
+	s.PrivateKey = conf.PrivateKey
+
+	if conf.PublicKey == "" {
+		return nil, errors.Errorf("genericrsa: missing public key for signer %q", s.ID)
+	}
+	s.key, s.pubKey, s.rng, s.PublicKey, err = conf.GetKeysAndRand()
+	if err != nil {
+		return nil, errors.Wrapf(err, "genericrsa: error fetching key and rand for signer %q", s.ID)
+	}
+	_, ok := s.pubKey.(*rsa.PublicKey)
+	if !ok {
+		return nil, errors.Errorf("genericrsa: unsupported public key type %T for signer %q, use RSA keys", s.pubKey, s.ID)
+	}
+
+	s.Hash = conf.Hash
+	var hRef crypto.Hash
```

Comment:
nit: this is colliding in my head with html `hrefs` maybe `hashID`

Reply:
fixed